n_i2c.c

```diff
@@ -212,72 +212,119 @@ const char *i2cNoteTransaction(char *request, char **response, size_t timeoutMs)
     delayIO();
-    const char *transmitErr = _I2CTransmit(_I2CAddress(), (uint8_t *)"\n", 1);
-    if (!cardTurboIO) {
-        _DelayMs(CARD_REQUEST_I2C_SEGMENT_DELAY_MS);
-    }
 
-    // This outer loop does retries on I2C error, and is simply here for robustness.
-    bool notecardReady = false;
-    int retries;
-    for (retries=0; transmitErr==NULL && !notecardReady && retries<3; retries++) {
+    // The guaranteed behavior for robust resyncing is to send two newlines
+    // and  wait for two echoed blank lines in return.
+    for (size_t retries = 0; retries < CARD_RESET_SYNC_RETRIES ; ++retries) {
+        // Send a newline to the module to clean out request/response processing
+        // NOTE: This MUST always be `\n` and not `\r\n`, because there are some
+        //       versions of the Notecard firmware will not respond to `\r\n`
+        //       after communicating over I2C.
+        const char *transmitErr = _I2CTransmit(_I2CAddress(), (uint8_t *)"\n", 1);
+        // If we get a failure on transmitting the `\n`, it means that the
+        // Notecard isn't present.
+        if (transmitErr) {
+            NOTE_C_LOG_ERROR(transmitErr);
+            NOTE_C_LOG_ERROR(ERRSTR("error encountered during I2C transmit hook execution", c_err));
+            _DelayMs(CARD_REQUEST_I2C_NACK_WAIT_MS);
+            notecardReady = false;
+            continue;
+        }
+
+        _DelayMs(CARD_REQUEST_I2C_SEGMENT_DELAY_MS);
 
-        // Loop to drain all chunks of data that may be ready to transmit to us
+        // Determine if I2C data is available
+        // set initial state of variable to perform query
         uint16_t chunkLen = 0;
-        while (true) {
+
+        // Content flags to determine if reset conditions are met.
+        bool somethingFound = false;
+        bool nonControlCharFound = false;
+
+        // Read I2C data for at least `CARD_RESET_DRAIN_MS` continously
+        for (const size_t startMs = _GetMs() ; (_GetMs() - startMs) < CARD_RESET_DRAIN_MS ;) {
 
             // Read the next chunk of available data
-            uint32_t available;
-            uint8_t buffer[128];
+            uint32_t available = 0;
+            uint8_t buffer[ALLOC_CHUNK] = {0};
             chunkLen = (chunkLen > sizeof(buffer)) ? sizeof(buffer) : chunkLen;
             chunkLen = (chunkLen > _I2CMax()) ? _I2CMax() : chunkLen;
             delayIO();
             const char *err = _I2CReceive(_I2CAddress(), buffer, chunkLen, &available);
             if (err) {
-                break;
+                // We have received a hardware or protocol level error.
+                // Introduce delay to relieve system stress.
+                NOTE_C_LOG_ERROR(err);
+                NOTE_C_LOG_ERROR(ERRSTR("error encountered during I2C receive hook execution", c_err));
+                _DelayMs(CARD_REQUEST_I2C_SEGMENT_DELAY_MS);
+                notecardReady = false;
+                continue;

```

Comment:
So when we fail to receive, we try to receive again, but we don't break out of this loop and try to transmit \n again first. I'm just putting words to it to make sure that's what we intend to do.

Reply:
We just want to clean the buffer. Theoretically, we are out-of-sync, and we just need to read anything back whether we fail or not.

Reply:
We will always resend `\n` if we were able to receive anything other than control characters in the 500 ms.